### PR TITLE
Implement pidfile Flock serialization

### PR DESCRIPTION
**Pull request by tempusfrangit**

### CODE REVIEW

File: cmd/root/root.go [hunk 1]
```diff
@@ -77,6 +96,18 @@ func GetCommand() *cobra.Command {
 	return cmd
 }
 
+// defaultPidFilePath returns the default path for the PID file. Notably modern OS X variants
+// do not have /var/run/
```
Comment (tempusfrangit) [inline cmd/root/root.go:100]:
```suggestion
// do not have /var/run or /run
```

File: pkg/cli/pid.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	"fmt"
+	"os"
+	"syscall"
+)
+
+type ProcessSerialization interface {
+	Acquire() error
+	Release() error
+}
```
Comment (tempusfrangit) [inline pkg/cli/pid.go:12]: This is useless... delete
```suggestion
```

File: cmd/root/root.go [hunk 3]
```diff
@@ -50,6 +50,25 @@ func GetCommand() *cobra.Command {
...
+			pidFilePath := viper.GetString(config.OptPIDFile)
+			pidFile, err := cli.NewPIDFile(pidFilePath)
+			if err != nil {
+				return err
+			}
+			err = pidFile.Acquire()
+			if err != nil {
+				return err
+			}
+			ctx := context.WithValue(cmd.Context(), config.OptPIDFile, pidFile)
```
Comment (philandstuff) [inline cmd/root/root.go:62]: I don't love using this constant for both the viper name and the context key.

also I think context keys should have user-defined types, not be strings?

Comment (philandstuff) [reply cmd/root/root.go:62]: actually I think I'd almost rather store this in a global variable at the top of this file, it feels like a lesser sin

File: cmd/root/root.go [hunk 4]
```diff
@@ -50,6 +50,25 @@ func GetCommand() *cobra.Command {
...
+			cmd.SetContext(ctx)
+			return nil
+		},
+		PersistentPostRunE: func(cmd *cobra.Command, args []string) error {
+			pidFile := cmd.Context().Value(config.OptPIDFile).(*cli.PIDFile)
+			err := pidFile.Release()
+			if err != nil {
+				return err
+			}
 			return nil
```
Comment (philandstuff) [inline cmd/root/root.go:72]:
```suggestion
			return := pidFile.Release()
```

File: cmd/root/root.go [hunk 5]
```diff
@@ -77,6 +96,18 @@ func GetCommand() *cobra.Command {
...
 }
 
+// defaultPidFilePath returns the default path for the PID file. Notably modern OS X variants
+// do not have /var/run/
+func defaultPidFilePath() string {
+	// If we're on OS X or Windows, use the user's home directory
+	// Otherwise, use /run
+	path := "/run/pget.pid"
+	if runtime.GOOS == "darwin" || runtime.GOOS == "windows" {
+		path = os.Getenv("HOME") + "/.pget.pid"
```
Comment (philandstuff) [inline cmd/root/root.go:106]: https://pkg.go.dev/os#UserHomeDir is slightly nicer I think, although you have to check for err. Or maybe even os.UserCacheDir?

File: pkg/cli/pid.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		func() error { return syscall.Flock(p.fd, syscall.LOCK_UN) },
+		p.file.Close,
+		func() error { return os.Remove(p.file.Name()) },
+	}
+	return p.executeFuncs(funcs)
+}
+
+func (p *PIDFile) writePID() error {
+	pid := os.Getpid()
+	_, err := p.file.Write([]byte(fmt.Sprintf("%d", pid)))
```
Comment (philandstuff) [inline pkg/cli/pid.go:47]:
```suggestion
	_, err := p.file.WriteString(fmt.Sprintf("%d", pid))
```
(double check this compiles but I think it's available here)

**Review by philandstuff — APPROVED** [review]
other than comments, this looks good

File: cmd/root/root.go [hunk 7]
```diff
@@ -77,6 +91,18 @@ func GetCommand() *cobra.Command {
 	return cmd
 }
 
+// defaultPidFilePath returns the default path for the PID file. Notably modern OS X variants
+// do not have /var/run or /run
```
Comment (tempusfrangit) [inline cmd/root/root.go:95]:
```suggestion
// have permissioning difficulties in /var/run etc.
```

**Pull request merged**